File: ch3/greedy/loowater_UVa11292.cpp
// The Dragon of Loowater

#include <bits/stdc++.h>
using namespace std;

typedef vector<int> vi;

int main() {
  int n, m;
  while (scanf("%d %d", &n, &m), (n || m)) {
    vi D(n);
    vi H(m);
    for (int d = 0; d < n; ++d) scanf("%d", &D[d]);
    for (int k = 0; k < m; ++k) scanf("%d", &H[k]);
    sort(D.begin(), D.end());                        // sorting is an important
    sort(H.begin(), H.end());                        // pre-processing step
    int gold = 0, d = 0, k = 0;                      // both arrays are sorted
    while ((d < n) && (k < m)) {                     // while not done yet
      while ((k < m) && (D[d] > H[k])) ++k;          // find required knight k
      if (k == m) break;                             // loowater is doomed :S
      gold += H[k];                                  // pay this amount of gold
      ++d; ++k;                                      // next dragon & knight
    }
    if (d == n) printf("%d\n", gold);                // all dragons are chopped
    else        printf("Loowater is doomed!\n");
  }
  return 0;
}

File: ch3/greedy/loowater_UVa11292.ml
open Printf
open Scanf

let () =
  try
    while true do
      let n, m = scanf "%d %d\n" (fun x y -> x, y) in
      if n = 0 && m = 0 then raise Exit;
      let dragon = Array.init n (fun _ -> scanf "%d\n" (fun x -> x)) in
      Array.sort compare dragon;
      let knight = Array.init m (fun _ -> scanf "%d\n" (fun x -> x)) in
      Array.sort compare knight;
      let gold = ref 0 in
      let d = ref 0 in
      let k = ref 0 in
      begin
        try
          while !d < n && !k < m do
            while !k < m && dragon.(!d) > knight.(!k) do k := !k+1 done;
            if !k = m then raise Exit;
            gold := !gold + knight.(!d);
            d := !d+1; k := !k+1;
          done
        with Exit -> ()
      end;
      if !d = n then printf "%d\n" !gold
      else printf "Loowater is doomed!\n"
    done
  with Exit -> ()
